---
editor_options:
  markdown:
    wrap: 72
output: pdf_document
---

```{r}
# Downloading the gene_expression.tsv file
download.file("https://raw.githubusercontent.com/ghazkha/Assessment4/refs/heads/main/gene_expression.tsv",
              "gene_expression.tsv")

# Install and load required packages
if (!require("seqinr")) install.packages("seqinr")

# Load seqinr package
library(seqinr)

# Read downloaded file
gene_expression <- read.table("gene_expression.tsv", header = TRUE, sep = "\t", row.names = 1)

# Display the first few rows of the gene expression data
head(gene_expression)

# Create a new column with the mean of the other columns
gene_expression$Mean <- rowMeans(gene_expression)

# Display the first few rows with the new Mean column
head(gene_expression)

# List the 10 genes with the highest mean expression
top_10_genes <- head(gene_expression[order(-gene_expression$Mean), ], 10)

# Display the top 10 genes
top_10_genes

# Count the number of genes with a mean expression less than 10
low_mean_genes <- sum(gene_expression$Mean < 10)

# Display the number of genes with mean expression < 10
low_mean_genes

```

```{r, echo=FALSE}
# Histogram of mean values
hist(gene_expression$Mean, main = "Histogram of Gene Expression Means", 
     xlab = "Mean Expression")

```
 
```{r}
# Downloading the growth_data.csv file
download.file("https://raw.githubusercontent.com/ghazkha/Assessment4/refs/heads/main/growth_data.csv",
              "growth_data.csv")

# Read and inspect the "growth_data.csv" file
growth_data <- read.csv("growth_data.csv")

# Display column names
colnames(growth_data)

# Calculate the mean and standard deviation of tree circumference at the start (2005) and end (2020) of the study for both sites
circumf_stats <- aggregate(cbind(Circumf_2005_cm, Circumf_2020_cm) ~ Site, 
                           data = growth_data, 
                           FUN = function(x) c(mean = mean(x), sd = sd(x)))
circumf_stats

# Boxplot of tree circumference in 2005 by site
boxplot(Circumf_2005_cm ~ Site, data = growth_data, 
        main = "Tree Circumference in 2005 by Site", 
        xlab = "Site", 
        ylab = "Circumference (cm)")

# Boxplot of tree circumference in 2020 by site
boxplot(Circumf_2020_cm ~ Site, data = growth_data, 
        main = "Tree Circumference in 2020 by Site", 
        xlab = "Site", 
        ylab = "Circumference (cm)")

# Calculate mean growth over the last 10 years (2010-2020) at each site
growth_data$Growth_10_years <- growth_data$Circumf_2020_cm - growth_data$Circumf_2010_cm
mean_growth <- aggregate(Growth_10_years ~ Site, data = growth_data, mean)
mean_growth

# Perform a t-test to estimate if the 10-year growth differs between the two sites
t_test_growth <- t.test(Growth_10_years ~ Site, data = growth_data)
t_test_growth$p.value


```

# Part 2

```{r}
# Part 2

# Install required packages
install.packages("R.utils")

# Load the necessary libraries
library(R.utils)

# URL for Saprospirales bacterium CDS
URL_saprospirales <- "https://ftp.ensemblgenomes.ebi.ac.uk/pub/bacteria/release-59/fasta/bacteria_58_collection/saprospirales_bacterium_gca_003448025/cds/Saprospirales_bacterium_gca_003448025.ASM344802v1.cds.all.fa.gz"

# Download the file
download.file(URL_saprospirales, destfile = "saprospirales_cds.fa.gz")

# Read the FASTA file using seqinr's read.fasta
saprospirales_cds <- read.fasta(file = "saprospirales_cds.fa.gz", seqtype = "DNA")
saprospirales_cds_count <- length(saprospirales_cds)



# URL for E. coli CDS
URL_ecoli <- "http://ftp.ensemblgenomes.org/pub/bacteria/release-53/fasta/bacteria_0_collection/escherichia_coli_str_k_12_substr_mg1655_gca_000005845/cds/Escherichia_coli_str_k_12_substr_mg1655_gca_000005845.ASM584v2.cds.all.fa.gz"

# Download the file
download.file(URL_ecoli, destfile = "ecoli_cds.fa.gz")

# Read the FASTA file for E. coli using seqinr's read.fasta
ecoli_cds <- read.fasta(file = "ecoli_cds.fa.gz", seqtype = "DNA")
ecoli_cds_count <- length(ecoli_cds)



# Create a table comparing the number of coding sequences for both organisms
cds_table <- data.frame(
  Organism = c("E. coli", "Saprospirales"),
  Coding_Sequences = c(ecoli_cds_count, saprospirales_cds_count)
)

print(cds_table)

```

## Difference between two organism:
Saprospirales bacterium has a higher number of coding sequences (4,527) compared to *E. coli* (4,239), indicating a potentially more complex genome and greater functional diversity. This difference suggests that *Saprospirales bacterium* may possess specialized genes that enable it to adapt to specific environmental conditions or ecological niches, whereas *E. coli* has a more streamlined genome optimized for versatility in various laboratory and natural settings.



```{r}

# Calculate the total coding DNA length for both organisms
ecoli_total_coding_dna <- sum(sapply(ecoli_cds, length))
sapros_total_coding_dna <- sum(sapply(saprospirales_cds, length))

# Create a table comparing the total coding DNA lengths
total_coding_table <- data.frame(
  Organism = c("E. coli", "Saprospirales"),
  Total_Coding_DNA = c(ecoli_total_coding_dna, sapros_total_coding_dna)
)

print(total_coding_table)
```

## Difference between two organism:
here


```{r}

# Calculate sequence lengths for both organisms
ecoli_cds_lengths <- sapply(ecoli_cds, length)
sapros_cds_lengths <- sapply(saprospirales_cds, length)

# Boxplot of coding sequence lengths
boxplot(ecoli_cds_lengths, sapros_cds_lengths,
        names = c("E. coli", "Saprospirales"),
        main = "Coding Sequence Lengths",
        ylab = "Sequence Length (bp)")

# Calculate the mean and median of sequence lengths for both organisms
mean_ecoli <- mean(ecoli_cds_lengths)
median_ecoli <- median(ecoli_cds_lengths)
mean_sapros <- mean(sapros_cds_lengths)
median_sapros <- median(sapros_cds_lengths)

cat("E. coli: Mean =", mean_ecoli, ", Median =", median_ecoli, "\n")
cat("Saprospirales: Mean =", mean_sapros, ", Median =", median_sapros, "\n")

```

## Difference between two organism:
here


```{r}

# Define the standard nucleotides
nucleotides <- c("A", "C", "G", "T")

# Combine all E. coli sequences for nucleotide frequency calculation
ecoli_concat <- paste(unlist(ecoli_cds), collapse = "")  # Concatenate all sequences
ecoli_nuc_freq <- table(factor(strsplit(ecoli_concat, split = "")[[1]], levels = nucleotides))

# Combine all Saprospirales sequences for nucleotide frequency calculation
sapros_concat <- paste(unlist(saprospirales_cds), collapse = "")  # Concatenate all sequences
sapros_nuc_freq <- table(factor(strsplit(sapros_concat, split = "")[[1]], levels = nucleotides))

# Create a matrix for the bar plot
nuc_freq_matrix <- rbind(as.numeric(ecoli_nuc_freq), as.numeric(sapros_nuc_freq))

# Create bar plots for nucleotide frequencies
barplot(nuc_freq_matrix, beside = TRUE,
        names.arg = nucleotides,
        legend.text = c("E. coli", "Saprospirales"),
        main = "Nucleotide Frequency Comparison",
        col = c("blue", "red"),
        ylab = "Frequency",
        args.legend = list(x = "topright"))
```

## Difference between two organism:
here



```{r}

# Calculate codon usage for E. coli
ecoli_codon_usage <- uco(unlist(ecoli_cds), frame = 0, index = "freq")

# Calculate codon usage for Saprospirales
sapros_codon_usage <- uco(unlist(saprospirales_cds), frame = 0, index = "freq")

# Check the structure of codon usage outputs
str(ecoli_codon_usage)
str(sapros_codon_usage)

# Extract the codon usage frequency values directly (uco might already be a vector)
ecoli_codon_usage_vector <- as.vector(ecoli_codon_usage)
sapros_codon_usage_vector <- as.vector(sapros_codon_usage)

# Ensure both vectors have the same length for comparison
if (length(ecoli_codon_usage_vector) == length(sapros_codon_usage_vector)) {
  
  # Create a matrix for codon usage comparison
  codon_usage_matrix <- rbind(ecoli_codon_usage_vector, sapros_codon_usage_vector)

  # Create a bar plot to compare codon usage bias
  barplot(codon_usage_matrix, beside = TRUE,
          main = "Codon Usage Bias",
          col = c("blue", "red"),
          ylab = "Codon Usage Frequency",
          legend.text = c("E. coli", "Saprospirales"),
          args.legend = list(x = "topright"))
  
} else {
  warning("Codon usage vectors for E. coli and Saprospirales have different lengths.")
}


```

## Difference between two organism:
here



